src/cluster_migrateslots.c

```diff
@@ -0,0 +1,1946 @@
+        if (job->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!isSlotMigrationJobFinished(job)) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+void clusterFailAllSlotExportsWithMessage(char *message) {
+    listNode *ln;
+    listIter li;
+    listRewind(server.cluster->slot_migration_jobs, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotMigrationJob *job = ln->value;
+        if (job->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!isSlotMigrationJobInProgress(job)) continue;
+        finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_FAILED, message);
+    }
+}
+
+size_t clusterGetTotalSlotExportBufferMemory(void) {
+    listNode *ln;
+    listIter li;
+    listRewind(server.cluster->slot_migration_jobs, &li);
+    size_t result = 0;
+    while ((ln = listNext(&li)) != NULL) {
+        slotMigrationJob *job = ln->value;
+        if (job->type != SLOT_MIGRATION_EXPORT || job->client == NULL) continue;
+        result += getClientOutputBufferMemoryUsage(job->client);
+    }
+    return result;
+}
+
+/* Create a slot export job with the given target and slot ranges. */
+slotMigrationJob *createSlotExportJob(clusterNode *target_node, list *slot_ranges) {
+    slotMigrationJob *job = zcalloc(sizeof(slotMigrationJob));
+
+    job->ctime = server.unixtime;
+    job->last_update = job->ctime;
+    job->last_ack = job->ctime;
+    job->type = SLOT_MIGRATION_EXPORT;
+    job->state = SLOT_EXPORT_CONNECTING;
+    job->slot_ranges = slot_ranges;
+    job->slot_ranges_str = representSlotRangeList(slot_ranges);
+    getRandomHexChars(job->name, sizeof(job->name));
+    memcpy(job->nodename, target_node->name, CLUSTER_NAMELEN);
+    job->description = generateSlotMigrationJobDescription(job, target_node);
+    return job;
+}
+
+/* Read a response to the SYNCSLOTS ESTABLISH response, accumulating it in a buffer, and moving to
+ * the next stage of the migration if the response is a success. If there is an error, fail the
+ * migration with the error message. */
+void slotMigrationJobReadEstablishResponse(connection *conn) {
+    client *c = (client *)connGetPrivateData(conn);
+    slotMigrationJob *job = c->slot_migration_job;
+    if (c->flag.close_asap || !isSlotMigrationJobInProgress(job)) {
+        return;
+    }
+    if (!job->response_buf) {
+        job->response_buf = sdsempty();
+        job->response_buf = sdsMakeRoomForNonGreedy(job->response_buf, PROTO_IOBUF_LEN);
+    }
+
+    int result;
+    result = connRead(conn, ((char *)job->response_buf) + sdslen(job->response_buf), sdsavail(job->response_buf));
+    if (result > 0) {
+        sdsIncrLen(job->response_buf, result);
+    }
+    if (conn->state != CONN_STATE_CONNECTED) {
+        freeClientAsync(c);
+        return;
+    }
+    if (sdslen(job->response_buf) >= sdsalloc(job->response_buf)) {
+        finishSlotMigrationJob(job, SLOT_MIGRATION_JOB_FAILED, "Response to establish job is larger than buffer limit");
+        return;
+    }
```

Comment:
I think sdslen == sdsalloc is ok as the null terminator is reserved. convert it to an assert?

```suggestion
serverAssert(sdslen(job->response_buf) <= sdsalloc(job->response_buf))
```

Reply:
My worry with asserting here is I don't want another node's response being too large to crash a different node. I.e. imagine someone is writing a tool like RIOT that tries to pretend to be Valkey in order to do some export, but they write a long response.

> sdslen == sdsalloc is ok

Maybe the >= check is confusing right now. But we are at most reading sdsavail in the connRead call. sdsalloc == sdslen + sdsavail. So the point being - if after this read, we have filled up the buffer (the sdslen == the entire sdsalloc), we should stop reading since we have no more space to read.

I think I can rewrite this a little clearer by saying sdsavail == 0